cdc/metadata/metadata.go

```diff
@@ -0,0 +1,217 @@
+//    with a given timeout, for example, 1s;
+// 2. On a capture, controller, owners and processors share one same Context, which is
+//    associated with deadline 10s. CaptureObservation.Heartbeat will refresh the deadline.
+// 3. Controller is binded with a lease (10+1+1)s, for deadline, heartbeat time-elapsed
+//    and network clock skew.
+// 4. Controller needs to consider re-schedule owners and processors from a capture,
+//    if the capture has been partitioned with metadata storage more than lease+5s;
+// ---------------------------------------------------------------- //
+
+// SchedState is the type of state to schedule owners and processors.
+type SchedState int
+
+const (
+	// SchedLaunched means the owner or processor is launched.
+	SchedLaunched SchedState = iota
+	// SchedStopping means the owner or processor is in stopping.
+	SchedStopping
+	// SchedStopped means the owner or processor is stopped.
+	SchedStopped
+)
+
+// ScheduledOwner is for owner and processor schedule.
+type ScheduledOwner struct {
+	*ChangefeedInfo
+	ChangefeedProgress
+	Capture *CaptureInfo
+	State   SchedState
+}
+
+// ScheduledProcessor is for owner and processor schedule.
+type ScheduledProcessor struct {
+	*ChangefeedInfo
+	ChangefeedProgress
+	Capture *CaptureInfo
+	State   SchedState
+}
+
+// ChangefeedSchedule is used to query changefeed schedule information.
+type ChangefeedSchedule struct {
+	Owner     ScheduledOwner
+	Processor []ScheduledProcessor
+}
+
+// CaptureObservation is for observing and updating metadata on a CAPTURE instance.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type CaptureObservation interface {
+	// CaptureInfo tells the caller who am I.
+	CaptureInfo() *CaptureInfo
+
+	// GetChangefeed queries one changefeed or all changefeeds.
+	GetChangefeeds(cf ...model.ChangeFeedID) ([]*ChangefeedInfo, error)
+
+	// GetCaptures queries one capture or all captures.
+	GetCaptures(cp ...string) ([]*CaptureInfo, error)
+
+	// Heartbeat tells the metadata storage I'm still alive.
+	Heartbeat(context.Context) error
+
+	// TakeControl blocks until becomes controller or gets canceled.
+	TakeControl() (ControllerObservation, error)
+
+	// Advance advances some changefeed progresses.
+	Advance(cfs []*ChangefeedInfo, progresses []*ChangefeedProgress) error
+
+	// Fetch the latest changefeed owner list from metadata.
+	RefreshOwners() <-chan []ScheduledOwner
+
+	// When an owner exits, inform the metadata storage.
+	PostOwnerStopped(cf *ChangefeedInfo) error
+
+	// Fetch the latest changefeed processors from metadata.
+	RefreshProcessors() <-chan []ScheduledProcessor
+
+	// When a processor exits, inform the metadata storage.
+	PostProcessorStopped(cf *ChangefeedInfo) error
+}
+
+// ControllerObservation is for observing and updating meta by Controller.
+//
+// All intrefaces are thread-safe and shares one same Context.
+type ControllerObservation interface {
```

Comment:
Need GC scheduler info

Reply:
GC should be handled by controller internally.